### ISSUE TITLE
Keras predict_on_batch() won't work on the first call

### ISSUE THREAD

**Issue by hsm207** [tensorflow/tensorflow#20089]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.8.0
- **Python version**: 3.6.5 |Anaconda custom (64-bit)| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)]
- **Bazel version (if compiling from source)**: NA
- **GCC/Compiler version (if compiling from source)**: NA
- **CUDA/cuDNN version**: NA
- **GPU model and memory**: NA
- **Exact command to reproduce**: see source code

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Calling `model.predict_on_batch(x)` the first time results in the following error:

```
Traceback (most recent call last):
  File "C:\Users\user\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py", line 2963, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-ae1592c2ea28>", line 22, in <module>
    model.predict_on_batch(x)
  File "C:\Users\user\Anaconda3\lib\site-packages\tensorflow\python\keras\_impl\keras\engine\training.py", line 1459, in predict_on_batch
    return self(inputs)  # pylint: disable=not-callable
  File "C:\Users\user\Anaconda3\lib\site-packages\tensorflow\python\keras\_impl\keras\engine\base_layer.py", line 314, in __call__
    output = super(Layer, self).__call__(inputs, *args, **kwargs)
  File "C:\Users\user\Anaconda3\lib\site-packages\tensorflow\python\layers\base.py", line 630, in __call__
    in_deferred_mode = isinstance(input_list[0], _DeferredTensor)
IndexError: list index out of range
```
However, executing the same line a second time gives the expected results.

### Source code / logs

```
import tensorflow as tf
import tensorflow.contrib.eager as tfe

tfe.enable_eager_execution()

class MyModel(tf.keras.Model):
    def __init__(self):
        super(MyModel, self).__init__()
        self.dense1 = tf.keras.layers.Dense(10)
        self.dense2 = tf.keras.layers.Dense(2, activation='softmax')

    def call(self, inputs, training=None, mask=None):
        results = self.dense1(inputs)
        results = self.dense2(results)

        return results


x = tf.random_uniform((100, 10))

model = MyModel()
model.predict_on_batch(x)
```

**Comment by Dref360** [comment]
Why do you want to use predict_on_batch?
Doing `model(x)` will work.

**Comment by asimshankar** [comment]
(BTW, this has been fixed and the fix will be available in the 1.9 release. In the mean time, as @Dref360 suggested, you could use `model(x)` in 1.8)
